src/Microsoft.Health.Fhir.Core/Features/Search/Expressions/Parsers/SearchValueExpressionBuilderHelper.cs

```diff
@@ -295,24 +296,58 @@ void ISearchValueVisitor.Visit(UriSearchValue uri)
         {
             EnsureArg.IsNotNull(uri, nameof(uri));
 
-            switch (_modifier?.SearchModifierCode)
+            // This if statement handles a URI search string input by the user, that looks like a canonical
+            if (uri.IsCanonical)
             {
-                case null:
-                    _outputExpression = Expression.StringEquals(FieldName.Uri, _componentIndex, uri.Uri, false);
-                    break;
-                case SearchModifierCode.Above:
-                    _outputExpression = Expression.And(
-                        Expression.EndsWith(FieldName.Uri, _componentIndex, uri.Uri, false),
-                        Expression.NotStartsWith(FieldName.Uri, _componentIndex, KnownUriSchemes.Urn, false));
-                    break;
-                case SearchModifierCode.Below:
-                    _outputExpression = Expression.And(
-                        Expression.StartsWith(FieldName.Uri, _componentIndex, uri.Uri, false),
-                        Expression.NotStartsWith(FieldName.Uri, _componentIndex, KnownUriSchemes.Urn, false));
-                    break;
-                default:
-                    ThrowModifierNotSupported();
-                    break;
+                var canonicalSearchExpressions = new List<Expression>();
+
+                canonicalSearchExpressions.Add(GetUriExpression(uri.Uri));
+
+                if (!string.IsNullOrEmpty(uri.Version))
+                {
+                    canonicalSearchExpressions.Add(Expression.StringEquals(FieldName.UriVersion, _componentIndex, uri.Version, false));
+                }
+
+                if (!string.IsNullOrEmpty(uri.Fragment))
+                {
+                    canonicalSearchExpressions.Add(Expression.StringEquals(FieldName.UriFragment, _componentIndex, uri.Fragment, false));
+                }
+
+                // Match on provided canonical components
+                var searchComponents = Expression.And(canonicalSearchExpressions.ToArray());
+
+                // Or match on entire URI when canonical components are not separated, this is the case when the model type is Uri or Url.
+                var matchFullUri = Expression.And(
+                    GetUriExpression(uri.ToString()),
+                    Expression.Missing(FieldName.UriVersion, _componentIndex),
+                    Expression.Missing(FieldName.UriFragment, _componentIndex));
+
+                _outputExpression = Expression.Or(matchFullUri, searchComponents);
```

Comment:
This query unfortunately leads to a table scan of the UriSearchParamTable.
```sql
      DECLARE @p0 SmallInt = 1214;
      DECLARE @p1 VarChar(256) = 'http://hl7.org/fhir/tests/StructureDefinition/test-lab-6c63d70b-c105-456e-a79b-186215fa25e3|1#fragment';
      DECLARE @p2 VarChar(256) = 'http://hl7.org/fhir/tests/StructureDefinition/test-lab-6c63d70b-c105-456e-a79b-186215fa25e3';
      DECLARE @p3 VarChar(64) = '1';
      DECLARE @p4 VarChar(256) = 'fragment';
      DECLARE @p5 SmallInt = 96;
      DECLARE @p6 Int = 11;

      SET STATISTICS IO ON;
      SET STATISTICS TIME ON;

      WITH cte0 AS
      (
          SELECT ResourceSurrogateId AS Sid1
          FROM dbo.UriSearchParam
          WHERE IsHistory = 0
              AND SearchParamId = @p0
              AND (Uri = @p1
              AND Fragment IS NULL
              AND Fragment IS NULL

              OR Uri = @p2
              AND Version = @p3
              AND Fragment = @p4
              )

              AND ResourceTypeId = @p5

      ),
      cte1 AS
      (
          SELECT DISTINCT TOP (@p6) Sid1, 1 AS IsMatch, 0 AS IsPartial
          FROM cte0
          ORDER BY  Sid1 ASC
      )
      SELECT DISTINCT r.ResourceTypeId, r.ResourceId, r.Version, r.IsDeleted, r.ResourceSurrogateId, r.RequestMethod, CAST(IsMatch AS bit) AS IsMatch, CAST(IsPartial AS bit) AS IsPartial, r.IsRawResourceMetaSet, r.RawResource
      FROM dbo.Resource r

      INNER JOIN cte1
      ON r.ResourceSurrogateId = cte1.Sid1
      ORDER BY r.ResourceSurrogateId ASC
      OPTION(RECOMPILE)
```
![image](https://user-images.githubusercontent.com/339432/105370127-417dfb80-5bd1-11eb-860f-c74fb72b9803.png)

If instead we generate a UNION instead of an OR, we will get a more efficient query.

Reply:
Or you can put an additional LIKE clause:
```
AND URI like 'http://hl7.org/fhir/tests/StructureDefinition/test-lab-2764a056-d766-4f6b-8ba2-734c861a2baev%'
AND (Uri = @p1
AND Fragment IS NULL
AND Fragment IS NULL
OR Uri = @p2
AND Version = @p3
AND Fragment = @p4)
```